python/lsst/daf/butler/server.py

```diff
@@ -0,0 +1,441 @@
+    DimensionConfig,
+    SerializedDataCoordinate,
+    SerializedDatasetRef,
+    SerializedDatasetType,
+    SerializedDimensionRecord,
+)
+from lsst.daf.butler.core.serverModels import (
+    ExpressionQueryParameter,
+    QueryDataIdsModel,
+    QueryDatasetsModel,
+    QueryDimensionRecordsModel,
+)
+from lsst.daf.butler.registry import CollectionType
+from pydantic import BaseModel
+
+BUTLER_ROOT = "ci_hsc_gen3/DATA"
+
+log = logging.getLogger("excalibur")
+
+
+class MaximalDataId(BaseModel):
+    """Something that looks like a DataId but isn't."""
+
+    detector: int | None = None
+    instrument: str | None = None
+    physical_filter: str | None = None
+    exposure: int | None = None
+
+
+class CollectionTypeNames(str, Enum):
+    """Collection type names supported by the interface."""
+
+    def _generate_next_value_(name, start, count, last_values) -> str:  # type: ignore  # noqa: N805
+        # Use the name directly as the value
+        return name
+
+    RUN = auto()
+    CALIBRATION = auto()
+    CHAINED = auto()
+    TAGGED = auto()
+
+
+app = FastAPI()
+app.add_middleware(GZipMiddleware, minimum_size=1000)
+
+
+GLOBAL_READWRITE_BUTLER = None
+GLOBAL_READONLY_BUTLER = None
+
+
+def _make_global_butler() -> None:
+    global GLOBAL_READONLY_BUTLER, GLOBAL_READWRITE_BUTLER
+    if GLOBAL_READONLY_BUTLER is None:
+        GLOBAL_READONLY_BUTLER = Butler(BUTLER_ROOT, writeable=False)
+    if GLOBAL_READWRITE_BUTLER is None:
+        GLOBAL_READWRITE_BUTLER = Butler(BUTLER_ROOT, writeable=True)
+
+
+def butler_readonly_dependency() -> Butler:
+    _make_global_butler()
+    return Butler(butler=GLOBAL_READONLY_BUTLER)
+
+
+def butler_readwrite_dependency() -> Butler:
+    _make_global_butler()
+    return Butler(butler=GLOBAL_READWRITE_BUTLER)
+
+
+def unpack_dataId(butler: Butler, data_id: SerializedDataCoordinate | None) -> DataCoordinate | None:
+    """Convert the serialized dataId back to full DataCoordinate.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The butler to use for registry and universe.
+    data_id : `SerializedDatacoordinate` or `None`
+        The serialized form.
+
+    Returns
+    -------
+    dataId : `DataCoordinate` or `None`
+        The DataId usable by registry.
+    """
+    if data_id is None:
+        return None
+    return DataCoordinate.from_simple(data_id, registry=butler.registry)
+
+
+@app.get("/butler/")
+def read_root() -> str:
+    return "Welcome to Excalibur... aka your Butler Server"
+
+
+@app.get("/butler/butler.json", response_model=dict[str, Any])
+def read_server_config() -> Mapping:
+    """Return the butler configuration that the client should use."""
+    config_str = f"""
+datastore:
+    root: {BUTLER_ROOT}
+registry:
+    cls: lsst.daf.butler.registries.remote.RemoteRegistry
+    db: <butlerRoot>
+"""
+    config = Config.fromString(config_str, format="yaml")
+    return config
+
+
+@app.get("/butler/universe", response_model=dict[str, Any])
```

Comment:
My feeling was that we would never need to change this and it's a fundamental part of the API that always has to work for all clients (no arguments, returns a dict). Maybe I'm wrong and I should add in a v1 as well here just in case.

Reply:
Added a v1 here as well.